File: services/identity-service/src/test/java/br/com/deltaglobalbank/identity/domain/user/UserTests.java
package br.com.deltaglobalbank.identity.domain.user;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.lenient;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.UUID;

import br.com.deltaglobalbank.identity.domain.shared.PasswordHasher;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.EnumSource;

class UserTests {

    private final LockoutPolicy policy = new LockoutPolicy(
        true,
        List.of(5, 3, 2),
        List.of(Duration.ofMinutes(5), Duration.ofMinutes(15))
    );

    @Nested
    class Authentication {

        @Test
        void mustAuthenticateSuccessfullyAndResetFailedAttemptsOnValidCredentials() {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Hashed Password"),
                UserStatus.ACTIVE, false, null, null, 3, null,
                Instant.now(), Instant.now()
            );

            PasswordHasher hasher = mock(PasswordHasher.class);
            when(hasher.matches("Password", new HashedPassword("Hashed Password"))).thenReturn(true);

            user.authenticate("Password", hasher, Instant.now(), policy);

            UserSnapshot snap = user.snapshot();
            Assertions.assertAll(
                () -> assertEquals(0, snap.failedAttempts()),
                () -> assertNull(snap.lockedUntil()),
                () -> assertNotNull(snap.lastLoginAt())
            );
        }

        @Test
        void mustThrowAndIncrementFailedAttemptsWhenPasswordDoesNotMatch() {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Hashed Password"),
                UserStatus.ACTIVE, false, null, null, 2, null,
                Instant.now(), Instant.now()
            );

            PasswordHasher hasher = mock(PasswordHasher.class);
            when(hasher.matches("wrong-password", new HashedPassword("Hashed Password"))).thenReturn(false);

            assertThrows(InvalidCredentialsException.class,
                () -> user.authenticate("wrong-password", hasher, Instant.now(), policy));

            assertEquals(3, user.snapshot().failedAttempts());
        }

        @ParameterizedTest
        @EnumSource(value = UserStatus.class, names = {"SUSPENDED"})
        void mustThrowWhenUserStatusIsNotActive(UserStatus status) {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Hashed Password"),
                status, false, null, null, 0, null,
                Instant.now(), Instant.now()
            );

            PasswordHasher hasher = mock(PasswordHasher.class);
            lenient().when(hasher.matches("Password", new HashedPassword("Hashed Password"))).thenReturn(true);

            assertThrows(UserNotActiveException.class,
                () -> user.authenticate("Password", hasher, Instant.now(), policy));
        }
    }

    @Nested
    class ChangePassword {

        @Test
        void mustUpdatePasswordHashAndClearMustChangeFlagOnSuccessfulChange() {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Hashed Old Password"),
                UserStatus.ACTIVE, true, null, null, 0, null,
                Instant.now(), Instant.now()
            );

            Password currentPassword = new Password("OldPassword1!");
            Password newPassword = new Password("NewPassword1!");

            PasswordHasher hasher = mock(PasswordHasher.class);
            when(hasher.matches(currentPassword.value(), new HashedPassword("Hashed Old Password"))).thenReturn(true);
            when(hasher.hash(newPassword)).thenReturn(new HashedPassword("New Hashed Password"));

            user.changePassword(currentPassword, newPassword, hasher);

            UserSnapshot snap = user.snapshot();
            Assertions.assertAll(
                () -> assertEquals(new HashedPassword("New Hashed Password"), snap.passwordHash()),
                () -> assertEquals(false, snap.mustChangePassword()),
                () -> assertNotNull(snap.passwordChangedAt())
            );
        }

        @Test
        void mustThrowWhenCurrentPasswordDoesNotMatch() {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Old Hashed Password"),
                UserStatus.ACTIVE, true, null, null, 0, null,
                Instant.now(), Instant.now()
            );

            Password currentPassword = new Password("OldPassword1!");
            Password newPassword = new Password("NewPassword1!");

            PasswordHasher hasher = mock(PasswordHasher.class);
            when(hasher.matches(currentPassword.value(), new HashedPassword("Old Hashed Password"))).thenReturn(false);

            assertThrows(CurrentPasswordIncorrectException.class,
                () -> user.changePassword(currentPassword, newPassword, hasher));

            assertEquals(new HashedPassword("Old Hashed Password"), user.snapshot().passwordHash());
        }

        @Test
        void mustThrowWhenNewPasswordEqualsCurrentPassword() {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Same Hashed Password"),
                UserStatus.ACTIVE, true, null, null, 0, null,
                Instant.now(), Instant.now()
            );

            Password samePassword = new Password("Password1!");

            PasswordHasher hasher = mock(PasswordHasher.class);
            when(hasher.matches(samePassword.value(), new HashedPassword("Same Hashed Password"))).thenReturn(true);

            assertThrows(NewPasswordSameAsCurrentException.class,
                () -> user.changePassword(samePassword, samePassword, hasher));

            assertEquals(new HashedPassword("Same Hashed Password"), user.snapshot().passwordHash());
        }

        @Test
        void mustThrowUserLockedWhenAccountIsPermanentlyLocked() {
            User user = new User(
                UUID.randomUUID(), UUID.randomUUID(), "New User",
                new Email("admin@delta.com"), new HashedPassword("Hashed Password"),
                UserStatus.LOCKED, false, null, null, 20, null,
                Instant.now(), Instant.now()
            );

            PasswordHasher hasher = mock(PasswordHasher.class);

            assertThrows(UserLockedException.class,
                () -> user.authenticate("Password", hasher, Instant.now(), policy));
        }
    }
}

File: services/identity-service/src/test/kotlin/br/com/deltaglobalbank/identity/domain/user/UserTests.kt
package br.com.deltaglobalbank.identity.domain.user

import br.com.deltaglobalbank.identity.domain.shared.PasswordHasher
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.every
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.assertAll
import org.junit.jupiter.api.assertNotNull
import org.junit.jupiter.api.assertNull
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.EnumSource
import java.time.Instant
import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals

@ExtendWith(MockKExtension::class)
class UserTests {

    @Nested
    inner class Authentication {
        @Test
        fun `must authenticate successfully and reset failed attempts on valid credentials`() {
            val user = User(
                id = UUID.randomUUID(),
                tenantId = UUID.randomUUID(),
                fullName = "New User",
                email = Email("admin@delta.com"),
                passwordHash = HashedPassword("Hashed Password"),
                status = UserStatus.ACTIVE,
                mustChangePassword = false,
                passwordChangedAt = null,
                lastLoginAt = null,
                failedAttempts = 3,
                lockedUntil = null,
                createdAt = Instant.now(),
                updatedAt = Instant.now()
            )

            val hasher = mockk<PasswordHasher>()
            every { hasher.matches("Password", HashedPassword("Hashed Password")) } returns true

            user.authenticate("Password", hasher)

            val snap = user.snapshot()
            assertAll(
                { assertEquals(0, snap.failedAttempts) },
                { assertNull(snap.lockedUntil) },
                { assertNotNull(snap.lastLoginAt) }

            )
        }

        @Test
        fun `must throw and increment failed attempts when password does not match`() {
            val user = User(
                id = UUID.randomUUID(),
                tenantId = UUID.randomUUID(),
                fullName = "New User",
                email = Email("admin@delta.com"),
                passwordHash = HashedPassword("Hashed Password"),
                status = UserStatus.ACTIVE,
                mustChangePassword = false,
                passwordChangedAt = null,
                lastLoginAt = null,
                failedAttempts = 2,
                lockedUntil = null,
                createdAt = Instant.now(),
                updatedAt = Instant.now()
            )

            val hasher = mockk<PasswordHasher>()
            every { hasher.matches("wrong-password", HashedPassword("Hashed Password")) } returns false

            assertThrows(InvalidCredentialsException::class.java) {
                user.authenticate("wrong-password", hasher)
            }

            assertEquals(3, user.snapshot().failedAttempts)
        }

        @ParameterizedTest
        @EnumSource(value = UserStatus::class, names = ["SUSPENDED", "LOCKED"])
        fun `must throw when user status is not active`(status: UserStatus) {
            val user = User(
                id = UUID.randomUUID(),
                tenantId = UUID.randomUUID(),
                fullName = "New User",
                email = Email("admin@delta.com"),
                passwordHash = HashedPassword("Hashed Password"),
                status = status,
                mustChangePassword = false,
                passwordChangedAt = null,
                lastLoginAt = null,
                failedAttempts = 0,
                lockedUntil = null,
                createdAt = Instant.now(),
                updatedAt = Instant.now()
            )

            val hasher = mockk<PasswordHasher>()
            every { hasher.matches("Password", HashedPassword("Hashed Password")) } returns true

            assertThrows(UserNotActiveException::class.java) {
                user.authenticate("Password", hasher)
            }
        }
    }

    @Nested
    inner class ChangePassword {
        @Test
        fun `must update password hash and clear must-change flag on successful change`(){
            val user = User(
                id = UUID.randomUUID(),
                tenantId = UUID.randomUUID(),
                fullName = "New User",
                email = Email("admin@delta.com"),
                passwordHash = HashedPassword("Hashed Old Password"),
                status = UserStatus.ACTIVE,
                mustChangePassword = true,
                passwordChangedAt = null,
                lastLoginAt = null,
                failedAttempts = 0,
                lockedUntil = null,
                createdAt = Instant.now(),
                updatedAt = Instant.now()
            )

            val currentPassword = Password("OldPassword1!")
            val newPassword = Password("NewPassword1!")

            val hasher = mockk<PasswordHasher>()
            every { hasher.matches(currentPassword.value, HashedPassword("Hashed Old Password")) } returns true
            every { hasher.hash(newPassword) } returns HashedPassword("New Hashed Password")

            user.changePassword(currentPassword, newPassword, hasher)

            val snap = user.snapshot()
            assertAll(
                { assertEquals(HashedPassword("New Hashed Password"), snap.passwordHash) },
                {assertEquals(false, snap.mustChangePassword)},
                { kotlin.test.assertNotNull(snap.passwordChangedAt)}
            )
        }

        @Test
        fun `must throw when current password does not match`(){
            val user = User(
                id = UUID.randomUUID(),
                tenantId = UUID.randomUUID(),
                fullName = "New User",
                email = Email("admin@delta.com"),
                passwordHash = HashedPassword("Old Hashed Password"),
                status = UserStatus.ACTIVE,
                mustChangePassword = true,
                passwordChangedAt = null,
                lastLoginAt = null,
                failedAttempts = 0,
                lockedUntil = null,
                createdAt = Instant.now(),
                updatedAt = Instant.now()
            )

            val currentPassword = Password("OldPassword1!")
            val newPassword = Password("NewPassword1!")

            val hasher = mockk<PasswordHasher>()
            every { hasher.matches(currentPassword.value, HashedPassword("Old Hashed Password")) } returns false

            assertThrows (CurrentPasswordIncorrectException::class.java) {
                user.changePassword(currentPassword, newPassword, hasher)
            }

            assertEquals(HashedPassword("Old Hashed Password"), user.snapshot().passwordHash)
        }

        @Test
        fun `must throw when new password equals current password`(){
            val user = User(
                id = UUID.randomUUID(),
                tenantId = UUID.randomUUID(),
                fullName = "New User",
                email = Email("admin@delta.com"),
                passwordHash = HashedPassword("Same Hashed Password"),
                status = UserStatus.ACTIVE,
                mustChangePassword = true,
                passwordChangedAt = null,
                lastLoginAt = null,
                failedAttempts = 0,
                lockedUntil = null,
                createdAt = Instant.now(),
                updatedAt = Instant.now()
            )

            val samePassword = Password("Password1!")

            val hasher = mockk<PasswordHasher>()
            every { hasher.matches(samePassword.value, HashedPassword("Same Hashed Password")) } returns true

            assertThrows(NewPasswordSameAsCurrentException::class.java) {
                user.changePassword(samePassword, samePassword, hasher)
            }

            assertEquals(HashedPassword("Same Hashed Password"), user.snapshot().passwordHash)
        }
    }

}